man/stack/sp_stack_push.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_STACK_PUSH 3 DATE "libstaple-VERSION"
.SH NAME
sp_stack_push,
sp_stack_pushc,
sp_stack_pushs,
sp_stack_pushi,
sp_stack_pushl,
sp_stack_pushsc,
sp_stack_pushuc,
sp_stack_pushus,
sp_stack_pushui,
sp_stack_pushul,
sp_stack_pushf,
sp_stack_pushd,
sp_stack_pushld,
sp_stack_pushll,
sp_stack_pushull,
sp_stack_pushu8,
sp_stack_pushu16,
sp_stack_pushu32,
sp_stack_pushu64,
sp_stack_pushi8,
sp_stack_pushi16,
sp_stack_pushi32,
sp_stack_pushi64,
sp_stack_pushb,
sp_stack_pushstr,
sp_stack_pushstrn
\- push a new element onto a stack
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_stack_push "(struct sp_stack"
.RI * stack ,
const void
.RI * elem )
.br
int
.BR sp_stack_pushc "(struct sp_stack
.RI * stack ,
char
.IR elem )
.br
int
.BR sp_stack_pushs "(struct sp_stack
.RI * stack ,
short
.IR elem )
.br
int
.BR sp_stack_pushi "(struct sp_stack
.RI * stack ,
int
.IR elem )
.br
int
.BR sp_stack_pushl "(struct sp_stack
.RI * stack ,
long
.IR elem )
.br
int
.BR sp_stack_pushsc "(struct sp_stack
.RI * stack ,
signed char
.IR elem )
.br
int
.BR sp_stack_pushuc "(struct sp_stack
.RI * stack ,
unsigned char
.IR elem )
.br
int
.BR sp_stack_pushus "(struct sp_stack
.RI * stack ,
unsigned short
.IR elem )
.br
int
.BR sp_stack_pushui "(struct sp_stack
.RI * stack ,
unsigned int
.IR elem )
.br
int
.BR sp_stack_pushul "(struct sp_stack
.RI * stack ,
unsigned long
.IR elem )
.br
int
.BR sp_stack_pushf "(struct sp_stack
.RI * stack ,
float
.IR elem )
.br
int
.BR sp_stack_pushd "(struct sp_stack
.RI * stack ,
double
.IR elem )
.br
int
.BR sp_stack_pushld "(struct sp_stack
.RI * stack ,
long double
.IR elem )
.br
int
.BR sp_stack_pushll "(struct sp_stack
.RI * stack ,
long long
.IR elem )
.br
int
.BR sp_stack_pushull "(struct sp_stack
.RI * stack ,
unsigned long long
.IR elem )
.br
int
.BR sp_stack_pushu8 "(struct sp_stack
.RI * stack ,
uint8_t
.IR elem )
.br
int
.BR sp_stack_pushu16 "(struct sp_stack
.RI * stack ,
uint16_t
.IR elem )
.br
int
.BR sp_stack_pushu32 "(struct sp_stack
.RI * stack ,
uint32_t
.IR elem )
.br
int
.BR sp_stack_pushu64 "(struct sp_stack
.RI * stack ,
uint64_t
.IR elem )
.br
int
.BR sp_stack_pushi8 "(struct sp_stack
.RI * stack ,
int8_t
.IR elem )
.br
int
.BR sp_stack_pushi16 "(struct sp_stack
.RI * stack ,
int16_t
.IR elem )
.br
int
.BR sp_stack_pushi32 "(struct sp_stack
.RI * stack ,
int32_t
.IR elem )
.br
int
.BR sp_stack_pushi64 "(struct sp_stack
.RI * stack ,
int64_t
.IR elem )
.br
int
.BR sp_stack_pushb "(struct sp_stack"
.RI * stack ,
int
.IR elem )
.br
int
.BR sp_stack_pushstr "(struct sp_stack"
.RI * stack ,
const char
.RI * elem )
.br
int
.BR sp_stack_pushstrn "(struct sp_stack"
.RI * stack ,
const char
.RI * elem ,
size_t
.IR len )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
The functions in the
.BR sp_stack_push ()
family are used to push a new element onto a stack. The element's size must
match the stack's
.I elem_size
field, so it is illegal to, for example, create a stack with
.I elem_size
equal to
.BR sizeof(int) ,
and then use it with the
.BR sp_stack_pushc ()
function.
.P
.I stack
is the stack that shall be pushed onto.
.P
.I
elem
is the element to be pushed.
.SS Generic Form
.I elem
must be a void pointer to the first byte of the element that shall be pushed.
The number of bytes to copy is equal to the
.IR stack->elem_size
field.
.SS Suffixed Form
.I elem
is passed directly in the function argument.
.P
The
.BR sp_stack_pushstr ()
function creates a copy of the string pointed to by
.I elem
and stores the new pointer on the stack. The string must fit in
.I SIZE_MAX
bytes, including the null terminator.
.P
The
.BR sp_stack_pushstrn ()
function is equivalent to
.BR sp_stack_pushstr (),
except it shall always copy exactly
.I len
bytes and append a null terminator at the end.
.I len
must be smaller than
.IR SIZE_MAX .
.SH RETURN VALUE
If successful, the functions in the
.BR sp_stack_push ()
family shall return 0. Otherwise, they shall return a code indicating the
error.
.SH ERRORS
The functions in the
.BR sp_stack_push ()
family shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I stack
is a null pointer (debug mode only).
.IP \fBSP_ERANGE\fP 1.5i
Maximum buffer size exceeded. See BUGS in
.BR sp_stack (7).
.IP \fBSP_ENOMEM\fP 1.5i
Memory allocation failed.
.SS Generic Form
The
.BR sp_stack_push ()
function shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I elem
is a null pointer (debug mode only).
.SS Suffixed Form
The suffixed
.BR sp_stack_push ()
functions shall fail if:
.IP \fBSP_EILLEGAL\fP 1.5i
.IR stack->elem_size
does not match the size of the
.I elem
argument (debug mode only).
.P
The
.BR sp_stack_pushstr ()
and
.BR sp_stack_pushstrn ()
functions shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I elem
is a null pointer (debug mode only).
.IP \fBSP_ERANGE\fP 1.5i
string too long or
.I len
too large.
.SH CONFORMING TO
See
.BR libstaple (7).
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_stack (7),
.BR sp_stack_create (3),
.BR sp_stack_destroy (3),
.BR sp_stack_clear (3),
.BR sp_stack_peek (3),
.BR sp_stack_pop (3),
.BR sp_stack_insert (3),
.BR sp_stack_remove (3),
.BR sp_stack_qinsert (3),
.BR sp_stack_qremove (3),
.BR sp_stack_get (3),
.BR sp_stack_set (3),
.BR sp_stack_eq (3),
.BR sp_stack_copy (3),
.BR sp_stack_map (3),
.BR sp_stack_print (3)


src/stack/sp_stack_push.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_stack.h"
#include "../internal.h"
#include "../sp_errcodes.h"
#include <string.h>

int sp_stack_push(struct sp_stack *stack, const void *elem)
{
#ifdef STAPLE_DEBUG
	if (stack == NULL) {
		error(("stack is NULL"));
		return SP_EINVAL;
	}
	if (elem == NULL) {
		error(("elem is NULL"));
		return SP_EINVAL;
	}
#endif
	if (sp_size_try_add(DATA_SIZE(stack), stack->elem_size))
		return SP_ERANGE;
	if (sp_buf_fit(&stack->data, stack->size, &stack->capacity, stack->elem_size))
		return SP_ENOMEM;
	memcpy((char*)stack->data + stack->size++ * stack->elem_size, elem, stack->elem_size);
	return 0;
}
